hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/EmptyContainerHandler.java

```diff
@@ -88,9 +88,9 @@ public boolean handle(ContainerCheckRequest request) {
   private boolean isContainerEmptyAndClosed(final ContainerInfo container,
       final Set<ContainerReplica> replicas) {
     return container.getState() == HddsProtos.LifeCycleState.CLOSED &&
-        container.getNumberOfKeys() == 0 && replicas.stream()
```

Comment:
Why remove the check for number of keys being zero?

Reply:
This is the intention of change, We don't want to depend on keyCount(which can be non zero even though container is empty) to determine whether container is empty or not, in some situation keyCount has gone wrong. Instead now we use empty flag for checking whether replica is empty or not based on the results from all replica, and that will judge container is empty or not.

Reply:
That makes sense but then we're not checking whether SCM expects a container to be empty on the basis of `ContainerInfo` (0 keys currently) against what the replicas are actually reporting. For example, consider an edge case where some replicas are empty and some are non empty because of arbitrary reasons. If the non empty replicas are lost, then `EmptyContainerHandler` will now mark this container as `EMPTY` on the basis of reports from other replicas. Whereas we probably want to call the container `MISSING`?

Reply:
This is a good point. Do we need to have an isEmpty flag on the containerInfo. This would be set to true when all the replicas report as empty. Right now, we are depending on the keyCount reported in the replicas, and when the count drops to zero in all replicas, then we update the keycount in containerInfo to zero. Perhaps we should update an empty flag in a similar way, when non of the replicas are reporting "isEmpty=false".

Reply:
> Do we need to have an isEmpty flag on the containerInfo. This would be set to true when all the replicas report as empty.

I agree with this suggestion.

Reply:
@siddhantsangwan, @sodonnel Thanks for pointing out. I have added flag for container info as well to determine whether all the replicas are empty. Can you please look again.

Reply:
> Do we need to have an isEmpty flag on the containerInfo. This would be set to true when all the replicas report as empty.

I don't think we should have this. We should only track each piece of information in one place. We track if all the replicas are empty based on the `isEmpty` flag of each replica. Now there would be a duplicate value that may not match the state of all the current replicas. The safest thing is to check the isEmpty flag on all replicas before deleting, making this extra flag redundant.

> For example, consider an edge case where some replicas are empty and some are non empty because of arbitrary reasons. If the non empty replicas are lost, then EmptyContainerHandler will now mark this container as EMPTY on the basis of reports from other replicas.

Having another `areAllEmpty` flag on the `ContainerInfo` does not change this case, because that flag is still set based on all the replicas that SCM sees. It is no different from checking all replicas at the time of container delete.